source/m2mdynamicmemory.cpp

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright (c) 2015 ARM Limited. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ * Licensed under the Apache License, Version 2.0 (the License); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdlib.h>
+#include "include/m2mdynamicmemory.h"
+#include "include/m2mdynmemLIB.h"
+
+int M2MDynamicMemory::memTotal = 0;
+int M2MDynamicMemory::memCount = 0;
+void * M2MDynamicMemory::heap = 0;
+void * M2MDynamicMemory::heapPtr = 0;
+int M2MDynamicMemory::referenceCount = 0;
+
+
+#define M2M_DYNMEM_LIB
+//#define M2M_PASSTHROUGH
+//#define M2M_TRACE_PRINTS
+
+#ifdef M2M_PASSTHROUGH
+#ifdef M2M_TRACE_PRINTS
+#include <malloc.h>
+#endif
+#endif
+
+#ifdef M2M_TRACE_PRINTS
+#include <stdio.h>
+#endif
+
+#ifdef M2M_DYNMEM_LIB
+static mem_stat_t memInfo;
+#endif
+static void memory_fail_callback(heap_fail_t fail) {
+#ifdef M2M_TRACE_PRINTS
+    printf("\nM2M memory failure: %u\n", fail);
```

Comment:
These can be considered temporary debug prints. I should study the mbed_trace system a bit before changing to use it.